test/e2e/framework/provider.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+type Factory func() (ProviderInterface, error)
+
+var (
+	providers = make(map[string]Factory)
+	mutex     sync.Mutex
+)
+
+// RegisterProvider is expected to be called during application init,
+// typically by an init function in a provider package.
+func RegisterProvider(name string, factory Factory) {
+	mutex.Lock()
+	defer mutex.Unlock()
+	if _, ok := providers[name]; ok {
+		panic(fmt.Sprintf("provider %s already registered", name))
+	}
+	providers[name] = factory
+}
+
+func init() {
+	// "local" can always be used.
+	RegisterProvider("local", func() (ProviderInterface, error) {
+		return NullProvider{}, nil
+	})
+	// The empty string also works, but triggers a warning.
+	RegisterProvider("", func() (ProviderInterface, error) {
+		Logf("The --provider flag is not set.  Treating as a conformance test.  Some tests may not be run.")
+		return NullProvider{}, nil
+	})
+}
+
+// SetupProviderConfig validates the chosen provider and creates
+// an interface instance for it.
+func SetupProviderConfig(providerName string) (ProviderInterface, error) {
+	var err error
+
+	mutex.Lock()
+	defer mutex.Unlock()
+	factory, ok := providers[providerName]
+	if !ok {
+		return nil, fmt.Errorf("The provider %s is unknown.", providerName)
+	}
+	provider, err := factory()
+
+	return provider, err
+}
+
+// ProviderInterface contains the implementation for certain
+// provider-specific functionality.
+type ProviderInterface interface {
+	FrameworkBeforeEach(f *Framework)
+	FrameworkAfterEach(f *Framework)
+
+	ResizeGroup(group string, size int32) error
+	GetGroupNodes(group string) ([]string, error)
+	GroupSize(group string) (int, error)
+
+	CreatePD(zone string) (string, error)
+	DeletePD(pdName string) error
+	CreatePVSource(zone, diskName string) (*v1.PersistentVolumeSource, error)
+	DeletePVSource(pvSource *v1.PersistentVolumeSource) error
```

Comment:
So umm yeah, we need to discuss this.

Reply:
SIG storage has already created a sub-framework that abstracts out volume plugin (and cloud provider) setup/teardown.  I don't think we need to add an additional interface here.  For storage, it is possible to test multiple volume plugins against a single cloud provider.

Reply:
The subframework defines an interface for different volume plugins to implement at test/e2e/storage/drivers/base.go.  I think many of these tests can be converted to use the new framework, it just needs an owner.

Reply:
`base.go` has the interface, `in_tree.go` the implementation. Looking at the implementation for GCE, I find:
```
func (g *gcePdDriver) CreateVolume(volType testpatterns.TestVolType) interface{} {
...
        By("creating a test gce pd volume")
        vname, err := framework.CreatePDWithRetry()
...
}
```

`framework.CreatePDWithRetry` is from `pv_util.go`. In other words, this new API just wraps the existing code. It cannot replace it. Even if the tests were rewritten to not call `CreatePDWithRetry` directly, the problem would still remain.

Either the new API must be implemented differently (without direct calls to specific implementations) or we keep it as-is and connect it with the actual implementation via the API in this PR.

Reply:
We could reimplement each driver's method to call the cloud provider's methods directly, and then put each driver implementation into its own package.  My main point was I don't think the Provider interface is the right place to put PD/Volume methods because a cloud provider can support multiple volume types.

Reply:
I agree that the "vendor" interface in this PR isn't ideal and something better could be designed and implemented - over time. Right now the interface closely resembles how the framework previously called vendor-specific code (`createPD` function became the `createPD` interface method). There's still time and opportunities to improve that interface while all affected code is in the same repo, but I'd like to keep that out of the current PR, otherwise it'll never get done.